File: linked_lists/swap_nodes_in_pairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# Iterative
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        dummy = cur = ListNode(-1)
        dummy.next = head

        while cur.next and cur.next.next:
            first = cur.next
            second = cur.next.next

            cur.next = second
            first.next = second.next
            second.next = first

            cur = cur.next.next

        return dummy.next


# Recursive
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head and head.next:
            nxt = head.next
            head.next = self.swapPairs(nxt.next)
            nxt.next = head
            return nxt

        return head

File: linked_lists/swap_nodes_in_pairs.rb
# Definition for singly-linked list.
class ListNode
    attr_accessor :val, :next

    def initialize(val)
        @val = val
        @next = nil
    end
end

# Iterative
# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
    dummy = ListNode.new(-1)
    dummy.next = head
    curr = dummy

    while curr.next && curr.next.next
        first, second = curr.next, curr.next.next

        curr.next = second
        first.next = second.next
        second.next = first

        curr = curr.next.next
    end

    dummy.next
end

# Recursive
# @param {ListNode} head
# @return {ListNode}
def swap_pairs(first)
    return first if !first || !first.next

    second = first.next
    first.next = swap_pairs(second.next)
    second.next = first

    second
end
